lib/iris/fileformats/grib/_save_rules.py

```diff
@@ -200,13 +199,23 @@ def shape_of_the_earth(cube, grib):
 
     # Spherical earth.
     if ellipsoid.inverse_flattening == 0.0:
-        gribapi.grib_set_long(grib, "shapeOfTheEarth", 1)
-        gribapi.grib_set_long(grib, "scaleFactorOfRadiusOfSphericalEarth", 0)
+        # Set shapeOfTheEarth key based on value of spherical earth radius.
+        if int(ellipsoid.semi_major_axis) == 6367470:
+            gribapi.grib_set_long(grib, "shapeOfTheEarth", 0)
+        elif int(ellipsoid.semi_major_axis) == 6371229:
+            gribapi.grib_set_long(grib, "shapeOfTheEarth", 6)
+        else:
+            gribapi.grib_set_long(grib, "shapeOfTheEarth", 1)
+        gribapi.grib_set_long(grib,
+                              "scaleFactorOfRadiusOfSphericalEarth", 0)
         gribapi.grib_set_long(grib, "scaledValueOfRadiusOfSphericalEarth",
                               ellipsoid.semi_major_axis)
     # Oblate spheroid earth.
     else:
-        gribapi.grib_set_long(grib, "shapeOfTheEarth", 7)
+        if isinstance(cs, iris.coord_systems.OSGB):
+            gribapi.grib_set_long(grib, "shapeOfTheEarth", 9)
+        else:
+            gribapi.grib_set_long(grib, "shapeOfTheEarth", 7)
```

Comment:
I think I left 3 out deliberately because separating 3 and 7 is conceptually quite hard as they are effectively identical keys with a difference of a factor of 1000 between them. If we currently support units of m then it seems extraneous to add or change the behaviour to expect km instead...